### PR TITLE
OSDOCS-5032 GA for Custom Metric Autoscaler 

### PR DESCRIPTION
**Pull request by mburke5678**
Despite the title, **this PR is for the TP 2.8.2 release, slated for Mar 6.** The title matches the related Jira doc and dev issues.

https://issues.redhat.com/browse/OSDOCS-5032

[Advanced features](https://issues.redhat.com/browse/OCPNODE-1252): 
* pausing replicas:  [Preview of New module](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-pausing_nodes-pods-autoscaling-custom)
* fallback Preview: [Step 1 code block, callout 11](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-creating-workload_nodes-pods-autoscaling-custom)
* customizable HPA name [Step 1 code block, callout 14](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-creating-workload_nodes-pods-autoscaling-custom)
* activation threshold Preview: [New third, fourth, and fifth paragraphs](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-about_nodes-pods-autoscaling-custom)

[audit logging ](https://issues.redhat.com/browse/OCPNODE-1394)  
* [Preview of new module](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-audit_nodes-pods-autoscaling-custom)

[CPU/Memory scaler](https://issues.redhat.com/browse/OCPNODE-1452)  

* [CPU Scaler preview new section](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-trigger-cpu_nodes-pods-autoscaling-custom)  
* [Memory Scaler preview new section](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-trigger-memory_nodes-pods-autoscaling-custom)

[Kafka Scaler](https://issues.redhat.com/browse/OCPNODE-1451) 
* [New section preview](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-trigger-kafka_nodes-pods-autoscaling-custom)

[Update the Prometheus scaler](https://keda.sh/docs/2.9/scalers/prometheus/#trigger-specification) 
* [Understanding the Prometheus trigger callout 8, 9, 10](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-trigger-prom_nodes-pods-autoscaling-custom)

[Release notes preview](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-rn_nodes-pods-autoscaling-custom)

### CODE REVIEW

File: modules/gathering-data-audit-logs.adoc [hunk 1]
```diff
@@ -6,9 +6,12 @@
 ifeval::["{context}" == "gathering-cluster-data"]
 :support:
 endif::[]
-ifeval::["{context}" == "audit-log-view"]
+ifeval::["{context}" == "audit-log-view"]
```
Comment (mburke5678) [inline modules/gathering-data-audit-logs.adoc:9]: This file is not used currently. We might need it later if/when Must Gather is added to custom metrics autoscaler

**Comment by mburke5678** [comment]
@joelsmith @gauron99 @zroubalik Can you PTAL at my PR?

File: modules/nodes-pods-autoscaling-custom-audit.adoc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      logOutputVolumeClaim: "persistentVolumeClaimName" <2>
+      policy:
+        rules: <3>
+        - level: Metadata
+        omitStages: "RequestReceived" <4>
+        omitManagedFields: false <5>
+      lifetime: <6>
+        maxAge: "2"
+        maxBackup: "1"
+        maxSi<ze: "50"
```
Comment (gauron99) [inline modules/nodes-pods-autoscaling-custom-audit.adoc:46]:
```suggestion
        maxSize: "50"
```

File: modules/nodes-pods-autoscaling-custom-about.adoc [hunk 3]
```diff
@@ -17,6 +17,12 @@ To use the custom metrics autoscaler, you create a `ScaledObject` or `ScaledJob`
 You can create only one scaled object or scaled job for each workload that you want to scale. Also, you cannot use a scaled object or scaled job and the horizontal pod autoscaler (HPA) on the same workload.
 ==== 
 
+The custom metrics autoscaler, unlike the HPA, can scale to zero. If you set the `minReplicas` value in the custom metrics autoscaler to `0`, the cluster metrics autoscaler scales the workload down from 1 to 0 replica to or up from 0 replicas to 1. This is known as the _activation phase_. After scaling up to 1 replica, the HPA takes control of the scaling. This is known as the _scaling phase_.
```
Comment (gauron99) [inline modules/nodes-pods-autoscaling-custom-about.adoc:20]:
```suggestion
The custom metrics autoscaler, unlike the HPA, can scale to zero. If you set the `minReplicas` value in the custom metrics autoscaler to `0`, the custom metrics autoscaler scales the workload down from 1 to 0 replicas or up from 0 replicas to 1. This is known as the _activation phase_. After scaling up to 1 replica, the HPA takes control of the scaling. This is known as the _scaling phase_. 
```

**Comment by gauron99** [comment]
looks good to me! just found a typo

File: modules/nodes-pods-autoscaling-custom-creating-workload.adoc [hunk 4]
```diff
@@ -42,75 +45,114 @@ Timestamp:             2019-05-23T18:47:56Z
...
+      logOutputVolumeClaim: "persistentVolumeClaimName" 
+      policy:
+        rules: 
+        - level: Metadata
+        omitStages: "RequestReceived" 
+        omitManagedFields: false 
+      lifetime: 
+        maxAge: "2"
+        maxBackup: "1"
+        maxSi<ze: "50"
```
Comment (gauron99) [inline modules/nodes-pods-autoscaling-custom-creating-workload.adoc:102]:
```suggestion
        maxSize: "50"
```

**Review by zroubalik — COMMENTED** [review]
Did a quick read, will do another proper review later.

File: modules/nodes-pods-autoscaling-custom-trigger.adoc [hunk 5]
```diff
@@ -13,14 +13,14 @@ Triggers, also known as scalers, provide the metrics that the Custom Metrics Aut
 The custom metrics autoscaler currently supports only the Prometheus trigger, which can use the installed {product-title} monitoring or an external Prometheus server as the metrics source.  
 ====
 
-//You can specify a single trigger or multiple triggers. When using multiple triggers, the scaling is based on the greatest value from all the triggers. This section contains examples of the triggers suported for use with {product-title}. 
+//You can specify a single trigger or multiple triggers. When using multiple triggers, the scaling is based on the greatest value from all the triggers. This section contains examples of the triggers supported for use with {product-title}.
```
[the remark this answers is not in the document]
Comment (mburke5678) [reply modules/nodes-pods-autoscaling-custom-trigger.adoc:16]: @zroubalik Thanks for the reminder; does this statement still apply in the 2.8.2 TP release version?

File: modules/nodes-pods-autoscaling-custom-trigger.adoc [hunk 6]
```diff
@@ -13,14 +13,14 @@ Triggers, also known as scalers, provide the metrics that the Custom Metrics Aut
 The custom metrics autoscaler currently supports only the Prometheus trigger, which can use the installed {product-title} monitoring or an external Prometheus server as the metrics source.
```
[the remark this answers is not in the document]
Comment (mburke5678) [reply modules/nodes-pods-autoscaling-custom-trigger.adoc:13]: Lame excuse department: My laptop crashed on my while editing these files. Apparently, I missed re-doing a few changes!

**Comment by mburke5678** [comment]
@joelsmith @nee1esh Can you PTAL at my draft of [release notes for Custom Metrics Autoscaler](https://github.com/openshift/openshift-docs/pull/55731/files#diff-3a25a4f473bbc3a7189a13ddaf596e8951ebfd476229fc593e5c2c6fe456e61c)? [Preview](https://55731--docspreview.netlify.app/openshift-enterprise/latest/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-rn_nodes-pods-autoscaling-custom)

**Comment by mburke5678** [comment]
@weinliu Can you PTAL at this PR for a QE review? Thank you in advance.

**Review by zroubalik — COMMENTED** [review]
@mburke5678 looking good!
Minor nits:

- the release notes section is missing Kafka scaler
- following `Additional resources` section link should be right below Prometheus scaler section, atm it is below Kafka scaler
img

**Comment by mburke5678** [comment]
@zroubalik Thanks for the additonal review. 

> the release notes section is missing Kafka scaler

I am under the impression that Kafka and the "Advanced features" are not officially in the 2.8.2 Tech Preview. At the least, last I heard, @weinliu is not testing Kafka or the advanced features until GA.

**Comment by mburke5678** [comment]
@zroubalik Thnak you for the additional review. 

>     the release notes section is missing Kafka scaler

I'm under the impression that Kafka (and the _Advanced features_  are not officially in the 2.8.2 TP release. Or, at least, @weinliu is testing them for GA.

>  following Additional resources section link should be right below Prometheus scaler section, atm it is below Kafka scaler

This is due to a limitation in the docs that don't allow us to add links to our module files. We add the link to the assembly file. All those trigger sections are in the same module. I plan for GA to split these trigger sections into separate modules for GA.

**Comment by weinliu** [comment]
> @zroubalik Thanks for the additonal review.
> 
> > the release notes section is missing Kafka scaler
> 
> I am under the impression that Kafka and the "Advanced features" are not officially in the 2.8.2 Tech Preview. At the least, last I heard, @weinliu is not testing Kafka or the advanced features until GA.

@mburke5678 Hi Mike, I'm afraid I did not make myslef clear last time.  The table is what I wanted to talk about,

Kafka is in 2.8.2 TP, but "Advanced features" are not.

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-227eb0d6-7fff-388b-44da-20ee31c817ab"><div dir="ltr" style="margin-left:0pt;" align="left">

Work Items | Summary | Fixed Version
-- | -- | --
OCPNODE-1366 (The parent) | GA support for "Custom Metric Autoscaler" | GA
OCPNODE-1231 | QE Support CPU/Memory Scaler | 2.8.2 TP
OCPNODE-1282 | CMA audit logs | 2.8.2 TP
OCPNODE-1230 | QE Support Kafka Scaler | 2.8.2 TP
OCPNODE-1233 | Test advanced features | GA
OCPNODE-1226 | Run automated testsuite as part of CMA build pipeline | GA

</div></b>Work Items
Summary
Fixed Version
[OCPNODE-1366](https://issues.redhat.com/browse/OCPNODE-1366) (The parent)
GA support for "Custom Metric Autoscaler"
GA
[OCPNODE-1231](https://issues.redhat.com/browse/OCPNODE-1231)
QE Support CPU/Memory Scaler
2.8.2 TP
[OCPNODE-1282](https://issues.redhat.com/browse/OCPNODE-1282)
CMA audit logs
2.8.2 TP
[OCPNODE-1230](https://issues.redhat.com/browse/OCPNODE-1230)
QE Support Kafka Scaler
2.8.2 TP
[OCPNODE-1233](https://issues.redhat.com/browse/OCPNODE-1233)
Test advanced features
GA
[OCPNODE-1226](https://issues.redhat.com/browse/OCPNODE-1226)
Run automated testsuite as part of CMA build pipeline
GA

File: modules/nodes-pods-autoscaling-custom-trigger.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (mburke5678) [reply modules/nodes-pods-autoscaling-custom-trigger.adoc:16]: @zroubalik ^^

[the remark this answers is not in the document]
Comment (zroubalik) [reply modules/nodes-pods-autoscaling-custom-trigger.adoc:16]: Yes!

[the remark this answers is not in the document]
Comment (zroubalik) [reply modules/nodes-pods-autoscaling-custom-trigger.adoc:16]: Sorry, missed the notifcation somehow

File: modules/nodes-pods-autoscaling-custom-trigger.adoc [hunk 7]
```diff
@@ -158,11 +135,74 @@ spec:
...
 <3> Specifies the value to trigger scaling actions for:
 * When using `Utilization`, the target value is the average of the resource metrics across all relevant pods, represented as a percentage of the requested value of the resource for the pods.
 * When using `AverageValue`, the target value is the average of the metrics across all relevant pods.
-////
+<4> Optional. Specifies an individual container to scale, based on the memory utilization of only that container, rather than the entire pod. Here, only the container named `api` is to be scaled.
+
+[id="nodes-pods-autoscaling-custom-trigger-kafka_{context}"]
+== Understanding the Kafka trigger
+
+You can scale pods based on an Apache Kafka topic or other service that supports the Kafka protocol. The custom metrics autoscaler does not scale higher than the number of Kafka partitions, unless you set the `allowIdleConsumers` parameter to `true` in the scaled object or scaled job.
```
Comment (nalhadef) [inline modules/nodes-pods-autoscaling-custom-trigger.adoc:150]: Should "other service" in the first line be "other services?"

File: modules/nodes-pods-autoscaling-custom-trigger.adoc [hunk 8]
```diff
@@ -158,11 +135,74 @@ spec:
...
+
+[NOTE]
+====
+If the number of consumer groups exceeds the number of partitions in a topic, the extra consumer groups sit idle.
+
+To avoid this, by default the number of replicas will not exceed:
```
Comment (nalhadef) [inline modules/nodes-pods-autoscaling-custom-trigger.adoc:156]: Avoid future talk by replacing "will not exceed' with. "does not exceed"

**Review by nalhadef — APPROVED** [review]

**Comment by nalhadef** [comment]
@mburke5678 , i put a couple of minor suggestions. Beyond that, it LGTM.

**Comment by weinliu** [comment]
Hi @mburke5678 
I checked the ones below,  all /lgtm

pausing replicas
fallback
customizable HPA name
activation threshold
CPU scaler
Memory scaler
Kafka scaler

**Pull request merged**